feat: add file_parser module and improve Axum path handling

modcrafter77 opened the pull request:
- Add new file_parser module with multi-format document parsing capabilities and modernize path handling to use Axum 0.8+ syntax throughout the framework.

File Parser Module:
- Implement unified document parsing for PDF, DOCX, HTML, and plain text
- Add intermediate representation (IR) for structured content extraction
- Support markdown conversion for parsed documents
- Provide REST API endpoints for file parsing operations

Path Handling Improvements:
- Add normalize_to_axum_path() and axum_to_openapi_path() helper functions
- Update OperationBuilder to store paths in Axum 0.8+ format ({id}, {*path})
- Update MIME validation middleware to use MatchedPath extension for accurate route matching

Middleware Enhancements:
- Implement per-operation MIME type validation with 415 responses
- Use Axum's MatchedPath extension for reliable parameter route matching
- Improve fallback handling for unmatched routes

modcrafter77 committed:
feat: add file_parser module and improve Axum path handling

- Add new file_parser module with multi-format document parsing capabilities
 and modernize path handling to use Axum 0.8+ syntax throughout the framework.

File Parser Module:
- Implement unified document parsing for PDF, DOCX, HTML, and plain text
- Add intermediate representation (IR) for structured content extraction
- Support markdown conversion for parsed documents
- Provide REST API endpoints for file parsing operations
- docx-rs crate is replaced with custom version with fix (issue will be addressed for official docx-rs crate)

Path Handling Improvements:
- Add normalize_to_axum_path() and axum_to_openapi_path() helper functions
- Update OperationBuilder to store paths in Axum 0.8+ format ({id}, {*path})
- Update MIME validation middleware to use MatchedPath extension for accurate route matching

Middleware Enhancements:
- Implement per-operation MIME type validation with 415 responses
- Use Axum's MatchedPath extension for reliable parameter route matching
- Improve fallback handling for unmatched routes
- All int16 codes are replaced with http::* constants
- add module name for Lifecycle (logging in stateful modules is improved

- E2E core infrastructure is implemented:
  - run in docker
  - run locally vs local hyperspot server

Artifizer commented on modules/file_parser/src/api/rest/dto.rs:
```diff
@@ -0,0 +1,182 @@
+
+/// REST DTO for parsed document metadata
+#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
+pub struct ParsedMetadataDto {
```
Can it be `ParsedDocMetadataDto` or `ParseFileMetadataDto`?

Artifizer commented on modules/file_parser/src/api/rest/dto.rs:
```diff
@@ -0,0 +1,182 @@
+
+/// REST DTO for parse local file request
+#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
+pub struct ParseLocalRequest {
```
Can it be ParseLocalFileRequest?

Artifizer commented on modules/file_parser/src/api/rest/handlers.rs:
```diff
@@ -0,0 +1,392 @@
+
+/// Upload and parse a file
+#[tracing::instrument(
+    name = "file_parser.upload",
```
`file_parser.upload` or `file_parser.upload_and_parse` ?

Artifizer commented on modules/file_parser/src/api/rest/routes.rs:
```diff
@@ -0,0 +1,173 @@
+        .summary("Upload and parse a file")
+        .description("Accepts multipart/form-data file uploads with field name 'file'")
+        .tag("File Parser")
+        .require_auth("file_parser", "write")
```
can we fail compilation (or any other sast) if .require_auth() is not used (forgotten)?
If auth is not needed, I'd prefer an empty call, but not missing require_auth()

modcrafter77 replied:
Yes, it's good idea. We can implement it via typestate machine like secure orm doing it. I suffest to do it in separate PR

modcrafter77 commented on modules/file_parser/src/api/rest/dto.rs:
```diff
@@ -0,0 +1,182 @@
+
+/// REST DTO for file parse response (with optional markdown)
+#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
+pub struct FileParseResponseDto {
```
Only Document-based naming will be used.

modcrafter77 commented on modules/file_parser/src/api/rest/dto.rs:
```diff
@@ -0,0 +1,182 @@
+    pub markdown: Option<String>,
+}
+
+// Conversion implementations
```
In this particilar module we have not to much converters but ok I move it to separate module mappers.rs

modcrafter77 commented on modules/file_parser/src/api/rest/routes.rs:
```diff
@@ -0,0 +1,173 @@
+            200,
+            "Parsed document with optional markdown",
+        )
+        .problem_response(openapi, 400, "Bad Request")
```
If you are about error codes we have .standard_errors(openapi) helper for all well-know errors.

Artifizer commented on modules/file_parser/src/api/rest/routes.rs:
```diff
@@ -0,0 +1,173 @@
+            "Parsed document with optional markdown",
+        )
+        .problem_response(openapi, 400, "Bad Request")
+        .problem_response(openapi, 401, "Unauthorized")
```
Can it be:
```.problem_response_x(openapi, http.status_code_401_unauthorized)```

Artifizer commented on modules/file_parser/src/api/rest/error.rs:
```diff
@@ -0,0 +1,43 @@
+        }
+
+        DomainError::UnsupportedFileType { extension } => Problem::new(
+            400,
```
Still int...

modcrafter77 committed:
refactor(errors):
- problem_response(openapi, http::StatusCode::XXX, ... ) is replaced with shortcuts errror_XXX()

modcrafter77 committed:
refactor(errors):
- problem_response(openapi, http::StatusCode::XXX, ... ) is replaced with shortcuts errror_XXX()

modcrafter77 committed:
Merge branch 'main' into feature/file_parser_module_latest
